machine-learning/02-openpbta_consensus_clustering.Rmd

```diff
@@ -0,0 +1,351 @@
+
+We used a different measure (VST values vs. FPKM) and used all features. 
+
+## Low-dimensional representation
+
+Let's see if low-dimensional representations agree with our other unsupervised results.
+First, let's find the high variance genes; we'll perform dimension reduction on those genes only.
+
+```{r gene_variance}
+# Calculate variance
+gene_variance <- matrixStats::rowVars(rnaseq_mat)
+# Find the value that we'll use as a threshold to filter the top 5%
+variance_threshold <- quantile(gene_variance, 0.95)
+# Row indices of high variance genes
+high_variance_index <- which(gene_variance > variance_threshold)
+```
+
+### UMAP
+
+First, let's use UMAP.
+
+```{r run_umap}
+# Set seed for reproducible UMAP results
+set.seed(2020)
+# umap() expects features (genes) to be columns, so we have to use t()
+umap_results <- umap::umap(t(rnaseq_mat[high_variance_index, ]))
+```
+
+The UMAP coordinates are in the `layout` element of the list returned by `umap::umap()`.
+
+```{r make_umap_df, live = TRUE}
+# Make a data frame of the layout results and join with molecular subtype 
+# data frame
+umap_plot_df <- data.frame(umap_results$layout) %>%
+  tibble::rownames_to_column("Kids_First_Biospecimen_ID") %>%
+  inner_join(mb_molecular_subtype_df)
+```
+
+Let's make a scatter plot and color our samples by the subtype labels.
+
+```{r umap_scatter}
+umap_plot_df %>%
+  ggplot(aes(x = X1, 
+             y = X2,
+             color = molecular_subtype)) +
+  geom_point() +
+  colorblindr::scale_color_OkabeIto() +
+  theme_bw() +
+  xlab("UMAP1") +
+  ylab("UMAP2")
+```
+
+### PCA 
+
+Now, we'll briefly show you how to use built-in functions for PCA on any matrix that's not in a specialized object for some kind of genomic data.
+
+```{r run_pca}
+# Like umap(), prcomp() expects the features to be columns
+pca_results <- prcomp(t(rnaseq_mat[high_variance_index, ]), 
+                      scale = TRUE)
+# The low-dimensional representation is returned in x
+pca_results$x[1:6, 1:6]
+```
+
+Let's get the first 10 PCs ready for plotting.
+
+```{r pca_df, live = TRUE}
+pca_plot_df <- data.frame(pca_results$x[, 1:10]) %>%
+  tibble::rownames_to_column("Kids_First_Biospecimen_ID") %>%
+  inner_join(mb_molecular_subtype_df)
+```
+
+And now make a scatterplot.
+
+```{r pca_scatter, live = TRUE}
+pca_plot_df %>%
+  ggplot(aes(x = PC1, 
+             y = PC2,
+             color = molecular_subtype)) +
+  geom_point() +
+  colorblindr::scale_color_OkabeIto() +
+  theme_bw() +
+  xlab("PC1") +
+  ylab("PC2")
+```
+
+[`plotly`](https://plotly.com/r/) is a package that allows us to plot interactive 3D scatterplots.
+
+```{r 3d_scatter}
+pca_3d_plot <- plot_ly() %>%
+  add_trace(data = pca_plot_df,
+            x = ~ PC1,
+            y = ~ PC2,
+            z = ~ PC3,
+            mode = "markers",
+            type = "scatter3d",
+            color = ~ molecular_subtype,
+            colors = colorblindr::palette_OkabeIto[1:4])
```

Comment:
If we are doing this, can we simplify it a bit? I'm not sure whether `add_trace()` is needed based on https://plotly.com/r/3d-scatter-plots/ It seems like we might be able to do this with something like the following, which is maybe more similar to ggplot syntax?
```
pca_3d_plot <- plot_ly(data = pca_plot_df,
                       x = ~ PC1,
                       y = ~ PC2,
                       z = ~ PC3,
                       color = ~ molecular_subtype,
                       colors = colorblindr::palette_OkabeIto[1:4],
                       mode = "markers",  # is this needed? not sure
                       type = "scatter3d"  # same here, does it automatically infer 3d? probably, but explicit is fine
                       ) %>%
  add_markers()
```

Reply:
Specifying `mode` and `type` avoids warnings that I thought could be potentially confusing (not sure how `add_markers()` fits in). I had trouble specifying `x`, etc. in the `plot_ly()` call but I think I was missing `~` when I tried that at first. I can take another look. 

This is probably not a good idea, but what about multiple 2D plots _and_ this 3D plot?

Reply:
TIMTOWTDI

I think `add_markers()` is more like `geom_point()` where `add_trace()` is a more general thing that can do lines too?